crates/etc-merge/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+
+    let mut new_etc_files = BTreeMap::new();
+    recurse_dir(new_etc, PathBuf::new(), &mut new_etc_files)
+        .context(format!("Recursing {new_etc:?}"))?;
+
+    let mut added = vec![];
+    let mut modified = vec![];
+
+    for (current_file, current_meta) in current_etc_files {
+        let Some(old_meta) = pristine_etc_files.get(&current_file) else {
+            // File was created
+            added.push(current_file);
+            continue;
+        };
+
+        if old_meta.metadata_hash != current_meta.metadata_hash
+            || old_meta.content_hash != current_meta.content_hash
+        {
+            modified.push(current_file.clone());
+        }
+
+        pristine_etc_files.remove(&current_file);
+    }
+
+    let removed = pristine_etc_files.into_keys().collect::<Vec<PathBuf>>();
+
+    Ok(Diff {
+        added,
+        modified,
+        removed,
+    })
+}
+
+fn compute_metadata_hash(meta: &cap_std::fs::Metadata) -> anyhow::Result<DigestBytes> {
+    let mut hasher = openssl::hash::Hasher::new(openssl::hash::MessageDigest::sha256())?;
+
+    let mut ty = vec![];
+
+    ty.push(meta.is_file() as u8);
+    ty.push(meta.is_dir() as u8);
+    ty.push(meta.is_symlink() as u8);
+
+    hasher.update(&ty)?;
+
+    if !meta.is_dir() {
+        hasher.update(&meta.len().to_le_bytes())?;
+    }
+
+    hasher.update(&meta.permissions().mode().to_le_bytes())?;
+
+    Ok(hasher.finish()?)
+}
+
+fn recurse_dir(dir: &CapStdDir, mut path: PathBuf, list: &mut Map) -> anyhow::Result<()> {
+    for entry in dir.entries()? {
+        let entry = entry.context(format!("Getting entry for {path:?}"))?;
+        let entry_name = entry.file_name();
+
+        path.push(&entry_name);
+
+        let entry_type = entry.file_type()?;
+
+        if entry_type.is_dir() {
+            let dir = dir
+                .open_dir(&entry_name)
+                .with_context(|| format!("Opening dir {path:?} inside {dir:?}"))?;
+
+            let metadata = dir
+                .metadata(".")
+                .context(format!("Getting dir meta for {path:?}"))?;
+
+            list.insert(
+                path.clone(),
+                Metadata::new("".into(), hex::encode(compute_metadata_hash(&metadata)?)),
+            );
+
+            recurse_dir(&dir, path.clone(), list).context(format!("Recursing {path:?}"))?;
+
+            path.pop();
+            continue;
+        }
+
+        let buf = if entry_type.is_symlink() {
+            let readlinkat_result = readlinkat(&dir, &entry_name, vec![])
+                .context(format!("readlinkat {entry_name:?}"))?;
+
+            readlinkat_result.into()
+        } else if entry_type.is_file() {
+            let mut buf = vec![0u8; entry.metadata()?.size() as usize];
+
+            let mut file = entry.open().context(format!("Opening entry {path:?}"))?;
+            file.read_exact(&mut buf)
+                .context(format!("Reading {path:?}. Buf: {buf:?}"))?;
```

Comment:
Two key things here. First I think we should incremental hashing instead of reading the whole thing into memory and then hashing the memory buffer.

But second a nice optimization to make here is handling the case where the file has fsverity enabled. Then we can efficiently just query its fsverity hash to do a diff.

It won't be the *default* for files in a mutable `/etc` to have fsverity enabled but it makes sense to do so (dir is mutable, so the write-tmpfile-then-rename pattern continues to work) and we should make things faster for those who do use it.